backend/src/order.ts

```diff
@@ -354,162 +310,173 @@ export function getAllForShow(show_id: number): Promise<IAdminOrderListItem[]> {
-          }
-          var url = body.url;
-          db.query('update nk2_orders set payment_url = :url where id = :order_id',
-            {order_id: order_id, url: url});
-
-          resolve({url: url});
-        });
-      });
-    })
-    .catch((err) => {
-      log.error('Failed to prepare payment', {error: err, order_id: order_id});
+      if (p === null) throw "No provider assigned to order";
+    } catch (err) {
+      log.error('Failed to get payment provider', {error: err, order_id});
       throw err;
-    });
+    }
+  }
+
+  return payment(p);
 }
 
-export function paymentCancelled(order_id: number, params): Promise<any> {
-  log.info('Payment was cancelled', {order_id: order_id});
+export async function preparePayment(order_id: number): Promise<any> {
+  let conn: IConnection;
+
+  log.info('Preparing payment', {order_id: order_id});
+  try {
+    conn = await db.beginTransaction();
+  } catch (err) {
+    log.error('Failed to prepare payment', {error: err, order_id});
+    throw err;
+  }
+
+  try {
+    log.log('debug', 'Get order payment status', {order_id});
+    const res = await db.query('select status from nk2_orders where id = :order_id', {order_id}, conn);
+    
+    log.log('debug', 'Reject if status is not seats-reserved', {order_id});
+    if (res[0]['status'] !== 'seats-reserved') {
+      const err = {name: "Validation error", message: "Invalid order status"}
+      throw err;
+    }
+
+    log.log('debug', 'Set order status to payment-pendig and add payment provider', {order_id});
+    await db.query('update nk2_orders set status = "payment-pending", payment_provider = :provider where id = :order_id', {order_id, provider: DEFAULT_PROVIDER}, conn);
 
-  var verification = [PAYTRAIL_PREFIX + order_id, params.TIMESTAMP, config.paytrail.password].join('|');
-  var verification_hash = md5(verification).toUpperCase();
+    log.log('debug', 'Commit payment status changes', {order_id});
+    await db.commit(conn);
 
-  if (verification_hash === params.RETURN_AUTHCODE) {
-    log.info('Verification hash matches - deleting order', {order_id: order_id});
-    return db.query('delete from nk2_orders where id = :order_id',
-      {order_id: order_id});
-  } else {
-    log.error('Hash verification failed!',
-      {order_id: order_id, verification_hash: verification_hash, return_authcode: params.RETURN_AUTHCODE});
-    throw 'Hash verification failed';
+  } catch (err) {
+    log.error('Failed to prepare payment', {error: err, order_id: order_id});
+    await db.rollback(conn);
+    throw err;
+  }
+
+  try {
+    const order = await get(order_id);
+    const args: ICreateArgs = {
+      successRedirect: config.public_url + 'api/orders/' + order_id + '/success',
+      errorRedirect: config.public_url + 'api/orders/' + order_id + '/failure',
+      successCallback: config.public_url + 'api/orders/' + order_id + '/notify/success',
+      errorCallback: config.public_url + 'api/orders/' + order_id + '/notify/failure',
+    };
```

Comment:
This actually looks like something that could be created just inside the payment handler's create() method. I'm fine with keeping this for now, but if you re-add support for paytrail you might want to get rid of this too.

Reply:
I'll think that I leave it now as it is. In future when support to paytrail is re-added or we want to support some other provider refactor of this will be neccessary.